src/storage/tracked.rs

```diff
@@ -0,0 +1,316 @@
+    /// This will compare the cache with the current storage and generate change events
+    /// in case the `PartialEq` implementation says that two components are different.
+    ///
+    /// If you don't care about `Change::Modified` events, you don't have to call this method.
+    ///
+    /// ## When should I call this method?
+    ///
+    /// You should make sure that it gets called before you need the information which component
+    /// has been modified. E.g. in case you have several systems writing to the component, then
+    /// several reading from it, you can just call `maintain_tracked` once in between.
+    pub fn maintain_tracked(&mut self)
+    where
+        T: PartialEq,
+    {
+        let (set, inner) = self.data.open_mut();
+        unsafe {
+            inner.maintain(set);
+        }
+    }
+
+    /// Resets the tracked storage. This clears all change events and the `changed` bitset.
+    /// You most likely want to do this at the end of every frame.
+    pub fn reset_tracked(&mut self) {
+        let (_, inner) = self.data.open_mut();
+        unsafe {
+            inner.reset();
+        }
+    }
+}
+
+#[derive(Derivative)]
+#[derivative(Default(bound = "S: Default"))]
+pub struct TrackedStorage<C, S = DenseVecStorage<C>> {
+    /// All `Inserted` and `Changed` components are marked.
+    changed: BitSet,
+    changes: Vec<Change>,
+    _marker: PhantomData<C>,
+    old: S,
+    storage: S,
+}
+
+impl<C, S> TrackedStorage<C, S>
+where
+    C: Clone,
+    S: UnprotectedStorage<C>,
+{
+    /// Returns a reference to the `changed` bitset,
+    /// which contains all components that have been inserted or modified
+    /// since the last `reset`.
+    pub fn changed(&self) -> &BitSet {
+        &self.changed
+    }
+
+    pub fn change_events<'a>(&'a self, e: &'a EntitiesRes) -> ChangeEvents<'a> {
+        let inner = e.join().zip(self.changes.iter());
+
+        ChangeEvents { inner }
+    }
+
+    unsafe fn reset(&mut self) {
+        for id in &self.changed {
+            let elem = self.old.get_mut(id);
+            *elem = self.storage.get(id).clone();
+        }
+
+        self.changed.clear();
+        self.changes.iter_mut().for_each(|c| *c = Change::None);
+    }
+
+    fn insert_change(changes: &mut Vec<Change>, id: Index, val: Change) {
+        use std::iter::repeat;
+
+        let ind = id as usize;
+        let len = changes.len();
+        changes.extend(repeat(Change::None).take((ind + 1) - len));
+
+        changes[ind].add(val);
+    }
+}
+
+impl<C, S> TrackedStorage<C, S>
+where
+    C: Clone + PartialEq,
+    S: UnprotectedStorage<C>,
+{
+    unsafe fn maintain(&mut self, set: &BitSet) {
+        let TrackedStorage {
+            ref old,
+            ref storage,
+            ref mut changes,
+            ..
+        } = *self;
+
+        set.iter()
+            .filter(|id| old.get(*id) != storage.get(*id))
+            .for_each(|id| Self::insert_change(changes, id, Change::Modified))
```

Comment:
This works, but also might get a bit expensive for large components.  We could probably make this more efficient by using something like this: [Dirty](https://bfrog.github.io/dirty/dirty/struct.Dirty.html) then implementing Component for it where T: Component, and using that as the internal component for TrackedStorage.  

From here we have two choices:

- Prioritize TrackedStorage efficiency, by assuming if the component is dirty then it was modified
- Prioritize TrackedStorage accuracy, by still checking the component against an old version but only when the component is dirty.

Reply:
Nice idea, but if you look at `get_mut`, you can see that we need to return a mutable reference to the component there, which means that once we iterate over components mutably we have a set dirty flag.

Reply:
And also this isn't expensive ^

Reply:
And for small components it will introduce unnecessary overhead.
I think user should have an option to choose if he wants accuracy or efficiency and if he wants to use `Dirty` wrapper.

Reply:
I only check equality for indices of `set` which only contains changed component indices.

Reply:
Which means this optimization is already implemented, just a little bit different

Reply:
Oh I meant to use `changes` here, not the `set` parameter. Sorry!

Reply:
I should probably go sleep because what I've been commenting here was nonsense.

Reply:
If my understanding is correct, then `changes` should contain an index for all components in question right?  Which means we're still doing this for all of them.  One thing I failed to mention though was that my idea with Dirty also included returning Dirty inside the Storage implementation, and using Dirty<T> as your component type.  So the end user who is iterating over the components will have to explicitly call `.write()` and set the dirty flag when they want to modify a component.

Reply:
Except my idea falls apart as soon as I remember that query specs by component type, not by storage type, so this wouldn't work.  My bad.

Reply:
I thought about that, but this is impossible without breakage: #168